File: python/tidegraph/frontier.py
"""Finite-input TimedDAG frontier relative to the region-block contract.

Potential event domains provide absence/completeness certificates. They do not
depend on route replay. General cycles use the streaming interpreter instead.
"""
from collections import defaultdict
from .blocks import canonicalize, deliver, evaluate_block
from .records import Result
from .validation import validate_window


def plan(graph, atoms, start, stop, max_events=1000000):
    order = graph.topological_order()
    domains = [set() for _ in graph.nodes]
    total = 0
    def add(v, batch, time):
        nonlocal total
        if start <= time < stop and (batch, time) not in domains[v]:
            domains[v].add((batch, time)); total += 1
            if total > max_events:
                raise ValueError("frontier potential-event limit exceeded; use streaming or a smaller window")
    for a in atoms:
        add(a.node, a.batch, a.time)
    offsets, edges = graph.adjacency()
    for v in order:
        for batch, time in domains[v]:
            for edge_id in edges[offsets[v]:offsets[v + 1]]:
                edge = graph.edges[edge_id]
                add(edge.target, batch, time + edge.delay)
    frames, dependencies = defaultdict(set), defaultdict(set)
    for v, domain in enumerate(domains):
        for batch, time in domain:
            frames[batch, graph.nodes[v].region, time].add(v)
    for edge in graph.edges:
        for batch, time in domains[edge.source]:
            target = batch, graph.nodes[edge.target].region, time + edge.delay
            if target in frames:
                dependencies[target].add((batch, graph.nodes[edge.source].region, time))
    return frames, dependencies


def run(graph, model, continuation, external, stop, *, sealed_until, mode="hard", zeta=1.0,
        trace=True, prefill=True, max_events=1000000):
    if mode not in {"hard", "hst", "softp"}:
        raise ValueError("invalid emit mode")
    inputs, ledger = validate_window(graph, model, continuation, external, stop, sealed_until)
    q = continuation.fork(); q.ledger = ledger
    atoms = list(q.pending) + inputs
    frames, dependencies = plan(graph, atoms, q.cut, stop, max_events)
    owners, fibers = defaultdict(list), defaultdict(list)
    for key in sorted(frames):
        owners[key[:2]].append(key)
    for a in atoms:
        fibers[a.batch, a.node, a.time].append(a)
    cursor = {owner: 0 for owner in owners}
    done, events, outputs, messages = set(), [], [], []
    stats = {"frontier_stages": 0, "region_blocks": 0, "state_blocks": 0, "state_steps": 0, "full_blocks": 0}
    while len(done) < len(frames):
        blocks = []
        for owner, keys in owners.items():
            block = []
            index = cursor[owner]
            while index < len(keys) and dependencies[keys[index]] <= done:
                block.append(keys[index]); index += 1
            if block:
                blocks.append((owner, block))
        if not blocks:
            raise RuntimeError("frontier failed to make progress")
        stats["frontier_stages"] += 1
        region_blocks = defaultdict(list)
        for owner, keys in blocks:
            region_blocks[owner[1]].extend(keys)
        for keys in region_blocks.values():
            ready = [(b, r, t, frames[b, r, t]) for b, r, t in sorted(keys, key=lambda k: (k[2], k[0]))]
            block_events, block_stats = evaluate_block(graph, model, q, ready, fibers, mode=mode, zeta=zeta, prefill=prefill)
            deliver(graph, model, block_events, fibers, messages, outputs)
            events.extend(block_events)
            for name, count in block_stats.items():
                stats[name] = stats.get(name, 0) + count
            stats["region_blocks"] += 1
        for owner, keys in blocks:
            cursor[owner] += len(keys); done.update(keys)
    q.pending = [a for bucket in fibers.values() for a in bucket if a.kind == 1 and a.time >= stop]
    q.cut = stop
    stats["candidate_events"] = len(events)
    return canonicalize(graph, Result(q, events if trace else [], outputs, messages if trace else [], stats))

File: cpp/src/frontier.cpp
#include "tide/frontier.h"
#include "tide/ops.h"
#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>

namespace tide {
Frontier::Frontier(Graph g, Model m, Options options)
    : graph_(std::move(g)), model_(std::move(m)), options_(options), pool_(options.workers) {
  graph_.compile(); graph_.topological_order(); validate_model(graph_, model_);
  if (options.mode != "hard" && options.mode != "hst" && options.mode != "softp") throw std::invalid_argument("invalid emit mode");
  if (options.max_events < 1 || !std::isfinite(options.zeta)) throw std::invalid_argument("invalid frontier options");
}
Result Frontier::run(const Continuation& initial, const std::vector<External>& external, Index stop, Index seal) {
  Result result;
  auto& q = result.continuation; q = initial;
  auto atoms = validate_window(graph_, model_, q, external, stop, seal);
  atoms.insert(atoms.end(), q.pending.begin(), q.pending.end());
  const auto frames = plan_frontier(graph_, atoms, q.cut, stop, options_.max_events);
  std::map<Owner, std::vector<Index>> owners;
  std::map<Owner, size_t> cursor;
  for (Index i = 0; i < static_cast<Index>(frames.size()); ++i) owners[{frames[i].batch, frames[i].region}].push_back(i);
  Fibers fibers;
  for (const auto& a : atoms) fibers[{a.batch, a.node, a.time}].push_back(a);
  std::set<Index> done;
  auto& stats = result.stats;
  stats = {{"frontier_stages", 0}, {"region_blocks", 0}, {"state_blocks", 0}, {"state_steps", 0}, {"full_blocks", 0}, {"candidate_events", 0}};
  while (done.size() < frames.size()) {
    std::map<Index, std::vector<Index>> blocks;
    for (const auto& [owner, ids] : owners) {
      auto index = cursor[owner];
      while (index < ids.size()) {
        const auto& deps = frames[ids[index]].dependencies;
        if (!std::includes(done.begin(), done.end(), deps.begin(), deps.end())) break;
        blocks[owner.second].push_back(ids[index]); ++index;
      }
    }
    if (blocks.empty()) throw std::runtime_error("frontier failed to make progress");
    ++stats["frontier_stages"];
    for (auto& [region, ids] : blocks) {
      std::sort(ids.begin(), ids.end(), [&](auto a, auto b) {
        return std::tie(frames[a].time, frames[a].batch) < std::tie(frames[b].time, frames[b].batch);
      });
      std::vector<Frame> ready;
      for (auto i : ids) ready.push_back(frames[i]);
      auto events = evaluate_block(graph_, model_, q, ready, fibers, options_, pool_, stats);
      ++stats["region_blocks"]; stats["candidate_events"] += events.size();
      for (auto& e : events) {
        if (e.active) {
          for (auto j = graph_.csr.offsets[e.node]; j < graph_.csr.offsets[e.node + 1]; ++j) {
            const auto id = graph_.csr.edges[j]; const auto& edge = graph_.edges[id];
            if (e.time > std::numeric_limits<Index>::max() - edge.delay) throw std::overflow_error("logical time overflow");
            Atom a{e.batch, edge.target, e.time + edge.delay, 1, id, e.time, e.full * model_.edge_scale[id]};
            fibers[{a.batch, a.node, a.time}].push_back(a);
            if (options_.trace) result.messages.push_back(a);
          }
          for (auto j = graph_.output_index.offsets[e.node]; j < graph_.output_index.offsets[e.node + 1]; ++j) {
            const auto port = graph_.output_index.edges[j];
            result.outputs.push_back({e.batch, e.time, port, e.full * model_.output_scale[port]});
          }
        }
        if (options_.trace) result.trace.push_back(std::move(e));
      }
      for (auto i : ids) { done.insert(i); ++cursor[{frames[i].batch, frames[i].region}]; }
    }
  }
  q.pending.clear();
  for (const auto& [coordinate, bucket] : fibers)
    for (const auto& a : bucket) if (a.kind == 1 && a.time >= stop) q.pending.push_back(a);
  q.cut = stop;
  std::sort(q.pending.begin(), q.pending.end(), [](const auto& a, const auto& b) { return a.key() < b.key(); });
  std::sort(result.trace.begin(), result.trace.end(), [](const auto& a, const auto& b) {
    return std::tie(a.time, a.batch, a.node) < std::tie(b.time, b.batch, b.node);
  });
  std::sort(result.messages.begin(), result.messages.end(), [&](const auto& a, const auto& b) {
    return std::tie(a.position, a.batch, graph_.edges[a.source].source, a.source)
         < std::tie(b.position, b.batch, graph_.edges[b.source].source, b.source);
  });
  std::sort(result.outputs.begin(), result.outputs.end(), [&](const auto& a, const auto& b) {
    return std::tie(a.time, a.batch, graph_.outputs[a.port], a.port) < std::tie(b.time, b.batch, graph_.outputs[b.port], b.port);
  });
  return result;
}
}  // namespace tide
